src/core/document.js

```diff
@@ -308,29 +309,40 @@ var Page = (function PageClosure() {
     },
 
     getAnnotationsData: function Page_getAnnotationsData(intent) {
-      var annotations = this.annotations;
-      var annotationsData = [];
-      for (var i = 0, n = annotations.length; i < n; ++i) {
-        if (!intent || isAnnotationRenderable(annotations[i], intent)) {
-          annotationsData.push(annotations[i].data);
+      var annotationsPromise = this.annotations ?
+        this.annotations : this.getAnnotations();
+      return annotationsPromise.then(function (annotations) {
+        var annotationsData = [];
+        for (var i = 0, n = annotations.length; i < n; ++i) {
+          if (!intent || isAnnotationRenderable(annotations[i], intent)) {
+            annotationsData.push(annotations[i].data);
+          }
         }
-      }
-      return annotationsData;
+        return annotationsData;
+      });
     },
 
-    get annotations() {
-      var annotations = [];
+    getAnnotations: function Page_getAnnotations() {
```

Comment:
Add the following right below this line:

```javascript
if (this._annotations) {
  return this._annotations;
}
```

This handles the caching for you in a private member variable. If `getAnnotations` is called for the first time, the code below this will be executed, but when it's called a second time, the cached variable will be returned.